daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/TransactionPreprocessor.scala

```diff
@@ -95,28 +61,52 @@ private[preprocessing] final class TransactionPreprocessor(
       tx: GenTransaction[NodeId, Cid]
   ): (ImmArray[speedy.Command], Set[ContractId]) = {
 
-    type Acc = ((Set[Value.ContractId], Set[Value.ContractId]), BackStack[speedy.Command])
+    type Acc = (Set[Value.ContractId], Set[Value.ContractId], BackStack[speedy.Command])
 
-    val ((localCids, _), cmds) =
-      tx.roots.foldLeft[Acc](((Set.empty, Set.empty), BackStack.empty)) {
-        case ((cids, stack), id) =>
-          tx.nodes.get(id) match {
+    val (_, globalCids, cmds) =
+      tx.roots.foldLeft[Acc]((Set.empty, Set.empty, BackStack.empty)) {
+        case ((localCids0, globalCids0, stack), id) =>
+          val (newGlobals, newLocals, cmd) = tx.nodes.get(id) match {
```

Comment:
I recommend keeping the order of global and local consistent, so use locals first here as well.

Reply:
:+1: